+sw_tests/+unit_tests/unittest_spinw_gencoupling.m

```diff
@@ -0,0 +1,195 @@
+            testCase.verify_val(testCase.default_coupling, ...
+                testCase.swobj.coupling)
+        end
+        
+        function test_gencoupling_with_maxSym_multiple_bonds(testCase)
+            testCase.swobj.genlattice('spgr', 'P 4')  % not overwrite abc
+            % set maxSym distance to only include bond idx = 1
+            testCase.swobj.gencoupling('maxDistance', 4.5 , 'maxSym', 4)
+            expected_coupling = testCase.default_coupling;
+            expected_coupling.dl = int32([1 0 1 1; 0 1 -1 1; 0 0 0 0]);
+            expected_coupling.atom1 = ones(1, 4, 'int32');
+            expected_coupling.atom2 = ones(1, 4, 'int32');
+            expected_coupling.mat_idx = zeros(3, 4, 'int32');
+            expected_coupling.idx = int32([1 1 2 2]);
+            expected_coupling.type = zeros(3, 4, 'int32');
+            expected_coupling.sym = zeros(3, 4, 'int32');
+            expected_coupling.nsym = int32(1);
+            testCase.verify_val(expected_coupling, testCase.swobj.coupling);
+            % increase maxSym to include bond idx = 2 (length sqrt(2)*3)
+            testCase.swobj.gencoupling('maxDistance', 4.5 , 'maxSym', 4.25);
+            expected_coupling.nsym = int32(2);
+            testCase.verify_val(expected_coupling, testCase.swobj.coupling);
+        end
+        
+        function test_gencoupling_maxSym_less_than_any_bond_length(testCase)
+            testCase.swobj.genlattice('spgr', 'P 4')  % not overwrite abc
+            testCase.verifyWarning(...
+                @() testCase.swobj.gencoupling('maxDistance', 4, ...
+                'maxSym', 2), 'spinw:gencoupling:maxSym')
+            % check nsym not set (bonds reduced to P0)
+            testCase.verify_val(testCase.default_coupling, ...
+                testCase.swobj.coupling)
+        end
+        
+        function test_gencoupling_when_tol_violates_symmetry(testCase)
+            testCase.swobj.genlattice('lat_const',[3 3.5 5], 'spgr', 'P 4')
+            % bond along a and b are sym. equiv. according to P 4
+            % but lengths differ by 0.5 Ang
+            testCase.verifyError(...
+                @() testCase.swobj.gencoupling('maxDistance', 3.25, ...
+                    'tolDist', 0.6), 'spinw:gencoupling:SymProblem')
+        end
+        
+        function test_gencoupling_with_two_mag_atoms(testCase)
+            testCase.swobj.addatom('r',[0.25 0.25 0.25],'S',2)
+            testCase.swobj.gencoupling('maxDistance', 3);
+            expected_coupling = testCase.default_coupling;
+            expected_coupling.dl = int32([0, 1, 0, 1, 0, 1, 0;
+                                          0, 0, 1, 0, 1, 0, 1;
+                                          0, 0, 0, 0, 0, 0, 0]);
+            expected_coupling.atom1 = int32([1 2 2 1 1 2 2]);
+            expected_coupling.atom2 = int32([2 1 1 1 1 2 2]);
+            expected_coupling.mat_idx = zeros(3, 7, 'int32');
+            expected_coupling.idx = int32([1 2 2 3 3 3 3]);
```

Comment:
@mducle in this test I have 2 magnetic ions and the bonds between `atom_1-atom_1`  and `atom_2-atom_2` in adjacent cells have the same bond index (`3`) - see table below for summary. As they are different species I would not have thought these bonds were symmetrically equivalent (despite having the same length) . Is this a bug?
```
	idx    subidx        dl                   dr               length      matom1      idx1      matom2      idx2
    ___    ______    ___________    _______________________    ______    __________    ____    __________    ____

     1       1       0    0    0     0.25     0.25     0.25    1.639     {'atom_1'}     1      {'atom_2'}     2  
     2       1       1    0    0     0.75    -0.25    -0.25    2.681     {'atom_2'}     2      {'atom_1'}     1  
     2       2       0    1    0    -0.25     0.75    -0.25    2.681     {'atom_2'}     2      {'atom_1'}     1  
     3       1       1    0    0        1        0        0        3     {'atom_1'}     1      {'atom_1'}     1  
     3       2       0    1    0        0        1        0        3     {'atom_1'}     1      {'atom_1'}     1  
     3       3       1    0    0        1        0        0        3     {'atom_2'}     2      {'atom_2'}     2  
     3       4       0    1    0        0        1        0        3     {'atom_2'}     2      {'atom_2'}     2  
```
it still happens even if you call `atom_2`->`Ce3+` (for example)

Reply:
So is the `spgr` specified for this case? The default setup in [line 18](https://github.com/SpinW/spinw/pull/89/files#diff-f31fd7d76f8d7f16c27a13c07286d5f2b5c2b83213148585fbcbed9c9f4fb7afR18) just specifies the lattice constant, and I thought in this case that the default behaviour is to sort bonds by length only...

Reply:
Correct there's no spacegroup provided in this example, but I expected bonds between the same pair of species to be labelled as symmetrically equivalent if they have the same bond-length (I would not have expected the  `atom_1-atom_1` bond to be equivalent to `atom_2-atom_2` as they are different species). But maybe I've misunderstood?

Reply:
Yeah, I see your point - even though we don't have any spacegroup information we still know that the atoms are different so the bonds _ought_ to be inequivalent. I guess the behaviour here is the result of the history of the code - I think the distance sorting was implemented first and then they symmetry checking was done on top of it... 

The question now is whether to change it - my concern is about backwards compatibility - I'll have a look at the documentation and tutorials to see if any examples rely on this behaviour...

Reply:
So [tutorial 31](https://spinw.org/tutorials/31tutorial) is the only published tutorial which relies on this behaviour - and it's just through sheer laziness in not specifying the atom labels (they should be the same type of atom). There are some other tutorials which are not published which rely on this too...

Now, we could do a couple of things:

1. Keep the current behaviour
2. Keep the current behaviour but emit a warning; maybe tell users to explicitly label their atoms if they haven't done so.
3. Change the behaviour to distinguish different atom labels _except_ the case of default labels (e.g. `atom_<x>`).
4. Change the behaviour to distinguish different atom labels in all cases and emit a warning stating the default behaviour has changed
5. Change the behaviour to distinguish different atom labels in all cases and stay silent.

What do you think? Can you think of any other options?

Reply:
We could do 2. for now and say that the behaviour will change in the future, create an issue and do 4. or 5. in a future release? I guess it depends if you consider this a bug, or a 'feature'...

It might be nice to have a 3.1.3 or 3.2 release of SpinW without any major changes except the slight improvements in warnings/consistency etc. that we're finding through the unit tests. We could discuss this in the Monday catch-up?

Reply:
As discussed offline, this bug will effect users (even if they have specified atom labels) so it is worth fixing. We've decided to go for option 4. The downside is this is a breaking change, although we can warn users about this in the CHANGELOG and in the code. It might break some systems tests etc. so depending on the size of the change and how much it breaks we may end up doing this in a separate PR.

Reply:
Ah sorry @rebeccafair  didn't see your previous comment!

Reply:
OK after talking to @rebeccafair lets discuss on Monday!